src/runtime/opencl/opencl_device_api.cc

```diff
@@ -529,9 +693,29 @@ void OpenCLWorkspace::Init(const std::string& type_key, const std::string& devic
     this->devices.insert(this->devices.end(), devices.begin(), devices.end());
     for (size_t i = 0; i < devices.size(); ++i) {
       cl_device_id did = devices[i];
-      device_to_platform[did] = platform;
+      CLDeviceInfo dev_info;
+      dev_info.platform_id = platform;
       this->queues.push_back(clCreateCommandQueue(this->contexts[platform], did, 0, &err_code));
       OPENCL_CHECK_ERROR(err_code);
+      cl_uint row_pitch;
+      OPENCL_CALL(clGetDeviceInfo(did, CL_DEVICE_IMAGE_PITCH_ALIGNMENT_KHR, sizeof(row_pitch),
+                                  &row_pitch, nullptr));
+      if (0 == row_pitch) {
+        row_pitch = kAllocAlignment;  // Fallback
+      }
+      dev_info.image_row_align = row_pitch;
+      size_t reqd_size = 0;
+      OPENCL_CALL(clGetDeviceInfo(did, CL_DEVICE_EXTENSIONS, 0, nullptr, &reqd_size));
+      std::vector<char> extn_buf(reqd_size);
+      OPENCL_CALL(clGetDeviceInfo(did, CL_DEVICE_EXTENSIONS, reqd_size, extn_buf.data(), nullptr));
+      std::string extensions(extn_buf.data());
+
+      if (extensions.find("cl_khr_image2d_from_buffer") != std::string::npos) {
+        dev_info.image_from_buffer_support = true;
+      } else {
+        dev_info.image_from_buffer_support = false;
+      }
```

Comment:
Just an idea of improvement. I suppose that this piece of code can be moved in a separate function, e.g. `bool IsOpenCLExtensionSupported(const std::string& name)` and this function can be used in different places with different extensions in the future.